src/mgr/MetricCollector.cc

```diff
@@ -22,30 +22,37 @@ MetricCollector<Query, Limit, Key, Report>::MetricCollector(MetricListener &list
 template <typename Query, typename Limit, typename Key, typename Report>
 MetricQueryID MetricCollector<Query, Limit, Key, Report>::add_query(
     const Query &query,
-    const std::optional<Limit> &limit) {
-  dout(20) << "query=" << query << ", limit=" << limit << dendl;
+    const std::optional<Limit> &limit,
+    const bool reregister) {
   uint64_t query_id;
   bool notify = false;
 
   {
     std::lock_guard locker(lock);
-
-    query_id = next_query_id++;
-    auto it = queries.find(query);
-    if (it == queries.end()) {
-      it = queries.emplace(query, std::map<MetricQueryID, OptionalLimit>{}).first;
-      notify = true;
-    } else if (is_limited(it->second)) {
-      notify = true;
+    if (reregister) {
+      dout(20) << "query=" << query << ", limit=" << (limit ? stringify(*limit) : "unlimited") << dendl;
+      auto it = queries.find(query);
```

Comment:
How about refactoring this as::

```
auto it = queries.find(query);
if (it == queries.end()) {
    query_id = next_query_id++;
    it = queries.emplace(query, std::map<MetricQueryID, OptionalLimit>{}).first;
    notify = true;
} else if (reregister || is_limited(it->second)) {
    notify = true;
}

// note that emplace does not do anything if the key already exists
it->second.emplace(query_id, limit);
counters.emplace(query_id, std::map<Key, PerformanceCounters>{});
```

?

Reply:
Then we need to find the query_id to be populated for the 'reregister' case. 

Your suggestion is good and I've considered this logic earlier, but then I dropped it not because of emplace but because the existing logic creates a new query_id (a new entry) everytime. But in the new reregister handler code, we just need to find something that's already existing in the map. Also I don't want to mess up with the 'is_limited' case, as the new query_id is being used for the 'is_limited' case too and then emplace it. So for the reregister handling, just finding the existing query_id and then do the handle_query_updated(). Hope this clarifies.

Reply:
OK. If you are concerned about `is_limited` part, you could add another `else {...}` block specific for re-register.